distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Naming_Properties.enso

```diff
@@ -0,0 +1,81 @@
+from Standard.Base import all
+import Standard.Base.Errors.Illegal_Argument.Illegal_Argument
+import Standard.Base.Errors.Illegal_State.Illegal_State
+import Standard.Base.Errors.Unimplemented.Unimplemented
+
+from project.Errors import Name_Too_Long
+
+## PRIVATE
+   An interface for a helper managing limitations of entity names.
+
+   It extends the API of the Java counterpart `NamingProperties`.
+type Naming_Properties
+    ## PRIVATE
+       Returns the size of the string in the target encoding.
+       It may throw a panic if `size_limit` is `Nothing`.
+    encoded_size : Text -> Integer ! Illegal_State
+    encoded_size name =
+        _ = name
+        Unimplemented.throw "This is an interface only."
+
+    ## PRIVATE
+       Truncates the string to the given size in the target encoding.
+       It may throw a panic if `size_limit` is `Nothing`.
+    truncate : Text -> Integer -> Text ! Illegal_State
+    truncate name size =
+        _ = [name, size]
+        Unimplemented.throw "This is an interface only."
+
+    ## PRIVATE
+       Returns the maximum size of an entity name in the target encoding,
+       or `Nothing` if there is no limit.
+    size_limit : Integer | Nothing
+    size_limit =
+        Unimplemented.throw "This is an interface only."
+
+    ## PRIVATE
+       Raises a `Name_Too_Long` error for the given name.
+       The implementation may customize the extra message appended to the error
+       explaining the limitations.
+    raise_name_too_long_error : Text -> Text -> Nothing ! Name_Too_Long
+    raise_name_too_long_error entity_kind name =
+        _ = [entity_kind, name]
+        Unimplemented.throw "This is an interface only."
+
+## PRIVATE
+   A default singleton implementation of a `Naming_Properties` which imposes no
+   length limit.
+type Unlimited_Naming_Properties
+    Instance
```

Comment:
Why have the Instance constructor?

Reply:
[I've tried making it a 'singleton type']() and it failed:

```
java.lang.UnsupportedOperationException: Unsupported operation identifier 'size_limit' and  object 'Unlimited_Naming_Properties'(language: Java, type: com.oracle.truffle.polyglot.PolyglotMap). Identifier is not executable or instantiable.
        at <Unknown Location> related to InterfaceProxy<class org.enso.interpreter.runtime.data.Type>
        at <Unknown Location> related to com.oracle.truffle.host.HostObject.doInvoke
        at <enso> Unique_Name_Strategy.type.new(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Table\0.0.0-dev\src\Internal\Unique_Name_Strategy.enso:43:24-61)
        at <enso> Unique_Naming_Strategy_Spec.spec<arg-1>(C:\NBO\enso\test\Table_Tests\src\Helpers\Unique_Naming_Strategy_Spec.enso:226:24-56)
        at <enso> Test.run_spec<arg-1>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:199:18-25)
        at <enso> Panic.catch(Internal)
        at <enso> Panic.type.recover(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Base\0.0.0-dev\src\Panic.enso:186-191)
        at <enso> Test.run_spec(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:198-202)
        at <enso> case_branch<arg-3>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:76:73-89)
        at <enso> case_branch<arg-1>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:76:49-90)
        at <enso> Runtime.no_inline(Internal)
        at <enso> Duration.type.time_execution(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Base\0.0.0-dev\src\Data\Time\Duration.enso:125:18-43)
        at <enso> case_branch(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:76:24-91)
        at <enso> Test.specify(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:75-77)
        at <enso> Unique_Naming_Strategy_Spec.spec<arg-1>(C:\NBO\enso\test\Table_Tests\src\Helpers\Unique_Naming_Strategy_Spec.enso:223-237)
        at <enso> case_branch<arg-1>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:39:29-37)
        at <enso> case_branch<arg-1>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:38-40)
        at <enso> Panic.catch(Internal)
        at <enso> case_branch(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:37-40)
        at <enso> Test.group<arg-1>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:30-45)
        at <enso> Test.group(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test.enso:29-45)
        at <enso> Unique_Naming_Strategy_Spec.spec(C:\NBO\enso\test\Table_Tests\src\Helpers\Unique_Naming_Strategy_Spec.enso:186-237)
        at <enso> Unique_Naming_Strategy_Spec.main<arg-1>(C:\NBO\enso\test\Table_Tests\src\Helpers\Unique_Naming_Strategy_Spec.enso:239:28-31)
        at <enso> Test_Suite.type.run<arg-1>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test_Suite.enso:69:17-21)
        at <enso> Test_Suite.type.run<arg-1>(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test_Suite.enso:68-70)
        at <enso> Test_Reporter.wrap_junit_testsuites(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test_Reporter.enso:18:14-19)
        at <enso> Test_Suite.type.run(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test_Suite.enso:67-70)
        at <enso> Test_Suite.type.run_main(C:\NBO\enso\built-distribution\enso-engine-0.0.0-dev-windows-amd64\enso-0.0.0-dev\lib\Standard\Test\0.0.0-dev\src\Test_Suite.enso:40:13-39)
        at <enso> Unique_Naming_Strategy_Spec.main(C:\NBO\enso\test\Table_Tests\src\Helpers\Unique_Naming_Strategy_Spec.enso:239:8-31)
```

I think an Enso 'type' with static methods behaves a bit differently from an instance with its member methods. Whether that is a good or bad thing, I'm not sure.

Reply:
To illustrate the difference, we can take a polyglot java code:
```java
package foo;

public class Foo {
    public static interface Fooable {
        public long foo();
    }

    public static long callFoo(Fooable f) {
        long x = f.foo();
        System.out.println("Foo.foo() returned " + x);
        return x;
    }
}
```

and try out in Enso:
```
from Standard.Base import all

polyglot java import foo.Foo

type Fooable_Class
    Instance x

    foo : Integer
    foo self = 100+self.x

type Fooable_Singleton
    foo : Integer
    foo = 200

main =
   IO.println (Foo.callFoo (Fooable_Class.Instance 42))
   IO.println (Foo.callFoo (Fooable_Singleton))
```

This currently yields:
```
Foo.foo() returned 142
142
Execution finished with an error: Unsupported operation identifier 'foo' and  object 'Fooable_Singleton'(language: Java, type: com.oracle.truffle.polyglot.PolyglotMap). Identifier is not executable or instantiable.
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotEngineException.unsupported(PolyglotEngineException.java:147)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotInteropErrors.invokeUnsupported(PolyglotInteropErrors.java:193)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotObjectProxyHandler$ProxyInvokeNode.invokeOrExecute(PolyglotObjectProxyHandler.java:261)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotObjectProxyHandler$ProxyInvokeNode.doCachedMethod(PolyglotObjectProxyHandler.java:214)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotObjectProxyHandlerFactory$ProxyInvokeNodeGen.executeAndSpecialize(PolyglotObjectProxyHandlerFactory.java:308)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotObjectProxyHandlerFactory$ProxyInvokeNodeGen.execute(PolyglotObjectProxyHandlerFactory.java:244)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotObjectProxyHandler$ObjectProxyNode.doDefault(PolyglotObjectProxyHandler.java:150)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotObjectProxyHandlerFactory$ObjectProxyNodeGen.executeAndSpecialize(PolyglotObjectProxyHandlerFactory.java:124)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotObjectProxyHandlerFactory$ObjectProxyNodeGen.executeImpl(PolyglotObjectProxyHandlerFactory.java:110)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.HostToGuestRootNode.execute(HostToGuestRootNode.java:124)
        at <java> org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotObjectProxyHandler.invoke(PolyglotObjectProxyHandler.java:105)
        at <java> jdk.proxy5/jdk.proxy5.$Proxy44.foo(Unknown Source)
        at <java> foo.Foo.callFoo(Foo.java:9)
        at <enso> Main.main<arg-1>(src\Main.enso:17:16-46)
        at <enso> Main.main(src\Main.enso:17:4-47)
```

Reply:
Indeed, an atom instance with member methods VS a type 'instance' with its static methods, in Enso look the same from the perspective of 'foo-ability', but in Java are different - an instance is happily proxied to handle the interface, whereas the type does not fit there.

Whether that is good or bad - IMO it depends on the intention.

IMO it may be that a type may behave differently from an instance, even if in Enso they look the same. 

@JaroslavTulach what do you think about this example? Does the current behaviour seem right or shall we try to make it more uniform with the Enso side where both approaches seem similar?